controller: Skip FPM truncation if TS is not enabled

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Fast partition movement works by setting the learner's raft start offset to
a safe prefix truncation of the local log, based on initial retention configs
and some calculations to determine an offset below which all data are recoverable
from cloud storage. Raft data above this offset are transferred directly,
whereas data below this offset (assumed to be recoverable from cloud) are not.

Prior to this change, this computation could produce an unsafe initial start
offset for the learner if archival_metadata_stm::max_removable_local_log_offset
exceeds the last uploaded offset, which can occur if archival is disabled.

Note that this difference between max_removable and last uploaded is by design
since automatic truncation is usually predicated on cleanup policy, and archival
being switched off should not block cleanup in the common case.

FPM is a special case in the sense that the "truncation" is not for cleanup
as such, but rather an omission of backed-up data from an expensive bulk
transfer. Which is to say it can and does occur even if the source topic is
compact-only.

The net result of this is that archival_metadata_stm reports that it is safe
to remove up to offset::max, and, as a result, FPM truncates the raft transfer
up to whatever removable offset is reported by the other stms. Any data produced
between the last cloud storage upload and this offset are lost.

This commit introduces checks to

- prevent FPM from adjusting the learner start offset above the start of the
  local log for any partition w/ TS disabled
- ensure that FPM  won't truncate past the last uploaded offset when TS is
  enabled but uploads are paused

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [x] v24.2.x

## Release Notes

### Bug Fixes

* Fixes a bug where data loss could occur during FPM w/ tiered storage disabled

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
controller: Skip FPM truncation if TS is not enabled

Fast partition movement works by setting the learner's raft start offset to
a safe prefix truncation of the local log, based on initial retention configs
and some calculations to determine an offset below which all data are recoverable
from cloud storage. Raft data above this offset are transferred directly,
whereas data below this offset (assumed to be recoverable from cloud) are not.

Prior to this change, this computation could produce an unsafe initial start
offset for the learner if archival_metadata_stm::max_removable_local_log_offset
exceeds the last uploaded offset, which can occur if archival is disabled.

Note that this difference between max_removable and last uploaded is by design
since automatic truncation is usually predicated on cleanup policy, and archival
being switched off should not block cleanup in the common case.

FPM is a special case in the sense that the "truncation" is not for cleanup
as such, but rather an omission of backed-up data from an expensive bulk
transfer. Which is to say it can and does occur even if the source topic is
compact-only.

The net result of this is that archival_metadata_stm reports that it is safe
to remove up to offset::max, and, as a result, FPM truncates the raft transfer
up to whatever removable offset is reported by the other stms. Any data produced
between the last cloud storage upload and this offset are lost.

This commit introduces checks to

- prevent FPM from adjusting the learner start offset above the start of the
  local log for any partition w/ TS disabled or w/ uploads paused
- prevent FPM from operating on internal topics
- ensure that FPM  won't truncate past the safe offset in general as reported
  by archival metadata stm. This is mostly a safeguard, as in most cases
  max_removable_local_log_offset should return the correct thing.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

bharathv commented on src/v/cluster/controller_backend.cc:
```diff
@@ -581,13 +590,16 @@ controller_backend::calculate_learner_initial_offset(
-      = p->max_removable_local_log_offset();
+    const auto max_removable_local_log_offset = std::min(
+      p->archival_meta_stm()->manifest().get_last_offset(),
+      p->max_removable_local_log_offset());
```
nit: probably helps to log the value of `p->max_removable_local_log_offset()` for debugging (helps see what STMs view of state) .

oleiman replied:
done

bharathv reviewed (commented):
looks great

WillemKauf commented on tests/rptest/tests/node_pool_migration_test.py:
```diff
@@ -397,3 +401,139 @@ def _quiescent_state():
+    @cluster(num_nodes=4,
+             log_allow_list=RESTART_LOG_ALLOW_LIST + TS_LOG_ALLOW_LIST)
+    def test_disable_tiered_storage(self):
+        '''
```
nit: missing comment?

oleiman replied:
added

WillemKauf commented on tests/rptest/tests/node_pool_migration_test.py:
```diff
@@ -397,3 +401,139 @@ def _quiescent_state():
+        second_start_offset = info.start_offset
+        second_hwm = info.high_watermark
+
+        # self.redpanda.rolling_restart_nodes(self.redpanda.nodes)
```
nit: dead code

oleiman replied:
removed

WillemKauf commented on src/v/cluster/controller_backend.cc:
```diff
@@ -581,13 +590,16 @@ controller_backend::calculate_learner_initial_offset(
-    const auto max_removable_local_log_offset
-      = p->max_removable_local_log_offset();
+    const auto max_removable_local_log_offset = std::min(
+      p->archival_meta_stm()->manifest().get_last_offset(),
```
In most worlds, `max_removable_local_log_offset()` should never be lower than `last_offset()`, can you add a quick comment explaining why this needs to be the `std::min()` between these two values for posterity?

WillemKauf reviewed (dismissed):
just nits

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68888-->
<!--87583d93-594d-4245-8c63-343a36ed37d9-->
<!--b4601c93-b587-4437-b414-b6b913604d7a-->
<!--d657b35c-00be-49e2-9216-3620a089be76-->
<!--23db8709-4e74-4c83-8980-7642f0476507-->
<!--ae0e76ff-caa2-4993-8f7c-28901d538677-->
<!--acb4d164-f919-4579-acac-a59ee33764ee-->
<!--789fcf5d-8624-429e-9f33-9686c58cdbe2-->
<!--3de5d9d0-719f-4338-9bed-b624345f2a97-->
<!--title-->
## Retry command for [Build#68888](https://buildkite.com/redpanda/redpanda/builds/68888)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/node_pool_migration_test.py::DisableTieredStorageTest.test_migrating_redpanda_nodes_to_new_pool@{"balancing_mode":"off","cleanup_policy":"compact,delete","test_mode":"tiered_storage_fast_moves"}
tests/rptest/tests/node_pool_migration_test.py::DisableTieredStorageTest.test_migrating_redpanda_nodes_to_new_pool@{"balancing_mode":"off","cleanup_policy":"compact","test_mode":"tiered_storage_fast_moves"}
tests/rptest/tests/node_pool_migration_test.py::DisableTieredStorageTest.test_migrating_redpanda_nodes_to_new_pool@{"balancing_mode":"node_add","cleanup_policy":"compact,delete","test_mode":"tiered_storage"}
tests/rptest/tests/node_pool_migration_test.py::DisableTieredStorageTest.test_migrating_redpanda_nodes_to_new_pool@{"balancing_mode":"off","cleanup_policy":"compact","test_mode":"no_tiered_storage"}
tests/rptest/tests/node_pool_migration_test.py::DisableTieredStorageTest.test_migrating_redpanda_nodes_to_new_pool@{"balancing_mode":"node_add","cleanup_policy":"compact","test_mode":"tiered_storage"}
tests/rptest/tests/node_pool_migration_test.py::DisableTieredStorageTest.test_migrating_redpanda_nodes_to_new_pool@{"balancing_mode":"node_add","cleanup_policy":"compact","test_mode":"no_tiered_storage"}
tests/rptest/tests/node_pool_migration_test.py::DisableTieredStorageTest.test_migrating_redpanda_nodes_to_new_pool@{"balancing_mode":"node_add","cleanup_policy":"compact,delete","test_mode":"no_tiered_storage"}
tests/rptest/tests/node_pool_migration_test.py::DisableTieredStorageTest.test_migrating_redpanda_nodes_to_new_pool@{"balancing_mode":"off","cleanup_policy":"compact,delete","test_mode":"no_tiered_storage"}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--83ac7e86-f20b-4775-9d1d-cbfc2bb745de-->
<!--TEST RESULTS-->
<!--9b6c61cf-e2c5-4422-8c23-e978b8d591c1-->
<!--TEST RESULTS-->
<!--a55ca731-05a9-4b1c-8d80-91dec5f0120f-->
<!--TEST RESULTS-->
<!--d81a18f1-b39f-4665-a155-7519df871870-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68888">build#68888</a></summary>


| test_class               | test_method                               | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "node_add", "cleanup_policy": "compact", "test_mode": "tiered_storage_fast_moves"}                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9765-45b9-bbfa-a4897abf1d0d | FAIL          | 0/19     |                                                                                                                                                                        |
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "node_add", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage_fast_moves"}                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9766-447a-8791-f217d92f6687 | FAIL          | 0/20     |                                                                                                                                                                        |
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "node_add", "cleanup_policy": "compact", "test_mode": "no_tiered_storage"}                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-975f-4e63-b3a9-84a61393a8aa | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "node_add", "cleanup_policy": "compact,delete", "test_mode": "no_tiered_storage"}                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9760-4023-aad5-0c34b5aae7dc | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "node_add", "cleanup_policy": "compact", "test_mode": "tiered_storage"}                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9761-42bd-8448-483b478c029b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "node_add", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage"}                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9762-432d-a931-e9bc8a782d44 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact", "test_mode": "tiered_storage_fast_moves"}                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9762-4196-b2cc-cd1329c7e925 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage_fast_moves"}                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9763-4b5a-a599-bb1d9aeb47c6 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact", "test_mode": "no_tiered_storage"}                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9764-4f4c-bd21-6757058bf684 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact,delete", "test_mode": "no_tiered_storage"}                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9765-43aa-91f7-7c0c20fcaa0a | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact", "test_mode": "tiered_storage"}                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9765-45b9-bbfa-a4897abf1d0d | FAIL          | 0/19     |                                                                                                                                                                        |
| DisableTieredStorageTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage"}                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9766-447a-8791-f217d92f6687 | FAIL          | 0/20     |                                                                                                                                                                        |
| RandomNodeOperationsTest | test_node_operations                      | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-9765-43aa-91f7-7c0c20fcaa0a | FLAKY         | 20/21    | upstream reliability is '99.53488372093024'. current run reliability is '95.23809523809523'. drift is 4.29679 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68895">build#68895</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68895#01980141-d6e6-41d1-826a-9e1fc2ef5a04 | FLAKY         | 20/21    | upstream reliability is '98.90510948905109'. current run reliability is '95.23809523809523'. drift is 3.66701 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68937">build#68937</a></summary>


| test_class                  | test_method                | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:---------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeE2ETests            | test_json_schema_unicode   | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68937#01980a75-9cd4-4446-8cc3-cfd9b7e7cfa4 | FLAKY         | 20/21    | upstream reliability is '97.67441860465115'. current run reliability is '95.23809523809523'. drift is 2.43632 and the allowed drift is set to 50. The test should PASS |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 1}                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68937#01980a75-9cd4-4446-8cc3-cfd9b7e7cfa4 | FLAKY         | 19/21    | upstream reliability is '97.4304068522484'. current run reliability is '90.47619047619048'. drift is 6.95422 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68951">build#68951</a></summary>


| test_class   | test_method                               | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-------------|:------------------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest | test_single_partition_leadership_transfer | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68951#01980af1-105e-43f1-b7b4-758087947227 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

oleiman commented:
CI Failure: https://buildkite.com/redpanda/redpanda/builds/68888#0197fcda-975f-4e63-b3a9-84a61393a8aa

Lazy Python Hack 🐍 🕳️

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/8df29601f3552dcf8438219fd58d186e5911f23d..4efa738b8cae414a46361e3d2c4cb6e83d3035ff) CR comments, minor test rework, and more words

oleiman commented on tests/rptest/tests/node_pool_migration_test.py:
```diff
@@ -397,3 +411,156 @@ def _quiescent_state():
+
+    @cluster(num_nodes=4,
+             log_allow_list=RESTART_LOG_ALLOW_LIST + TS_LOG_ALLOW_LIST)
+    def test_disable_tiered_storage(self):
```
i think we should add a case for upload pause

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/4efa738b8cae414a46361e3d2c4cb6e83d3035ff..e287125086ed4c7c75ed0e2648305f088c4268e9) test case for pause uploads

oleiman commented:
/ci-repeat 3
skip-redpanda-build
skip-units
tests/rptest/tests/node_pool_migration_test.py::DisableTieredStorageTest
dt-repeat=30

mmaslankaprv commented on src/v/cluster/controller_backend.cc:
```diff
@@ -581,20 +590,37 @@ controller_backend::calculate_learner_initial_offset(
-    const auto max_removable_local_log_offset
-      = p->max_removable_local_log_offset();
+    auto max_removable_local_log_offset = p->max_removable_local_log_offset();
+    auto last_uploaded = p->archival_meta_stm()->manifest().get_last_offset();
```
I am not sure if this is safe, there are multiple offsets exposed by `archival_metadata_stm` are we certain that this one is the right one to use ?

mmaslankaprv replied:
There is also `get_archive_clean_offset`

oleiman replied:
`last_offset` should be safe in the sense that it implies a successful upload and replicated add_segment command.

but yeah, aiui clean offset is the offset of the last command batch  applied to the last manifest that was synced to s3, so i guess it could lag. @Lazin is that possible? 

if so, i think we want `min(last, clean)`

Lazin replied:
The `get_archive_clean_offset` is not the right offset. It's used by the spillover mechanism. The 'archive' in the name refers to the part of the manifest which spilled over.

The `last_offset` is also wrong offset to use. This is how the state looks normally:

```
                                
                                 
 ┌───────┐ ┌────────────┐        
 │cloud  │ │ cloud      │        
 │log    │ │ manifest   │        
 └───────┘ └─────┬──────┘        
     ▲      clean│offset         
     │           ▼               
 ┌───┴───┬────────────────┐      
 │ local │                │      
 │ log   │                │      
 └───────┴────────────────┘      
         ▲              ▲        
         │              │        
     last offset  insync offset  
                                 
                                
```

The `last_offset` points to the last offset that was backed up to S3 but it may not be recoverable yet. This offset is stored in the local `archival_metadata_stm` snapshot. This offset usually lags behind quite a lot and it's not guaranteed that it is recoverable at all. What happens is:
- `ntp_archiver` uploads a segment
- it replicates command with metadata which advances the `last_offset`
- the command gets replicated and applied to the in -memory state of the STM (advancing `insync_offset`)
- eventually the in-memory state of the STM gets uploaded to S3 in form of the manifest and this when the `clean_offset` is advanced. The `clean_offset` is an `insync_offset` of the state which was successfully uploaded to S3 last.

So the `clean_offset` is the recoverable offset. If we download the manifest it will contain the state of the archival STM which has all commands in the log up to `clean_offset` applied to it. And the `clean_offset` can lag behind if we fail to upload the manifest in time (if we're getting throttled or something like that). In some situations the `ntp_archiver` allows 30s delay before the manifest is even uploaded. And the manifest uploads can be throttled too. So it is quite possible that `clean_offset` is below `last_offset`.

```
 ┌────────────┐                     
 │ cloud      │                     
 │ manifest   │                     
 └─────┬──────┘                     
  clean│offset                      
       ▼                            
   ┌───────┬────────────────┐       
   │ local │                │       
   │ log   │                │       
   └───────┴────────────────┘       
           ▲              ▲         
           │              │         
       last offset  insync offset                                  
```

If this is the case if we use `last_offset` as a truncation point for FPM we will end up in a situation when after the move the STM can't advance forward because there is a gap between the first offset in the local log and the insync offset of the STM.

In the `archival_metadata_stm` there is a method called `max_removable_local_log_offset` which performs these offset calculations. It takes into account some corner cases (like RRR) but after that what it does is essentially this:

```
return std::min(archival_metadata_stm.get_last_offset(), 
                archival_metadata_stm.get_last_clean_at());
```

Lazin commented on src/v/cluster/controller_backend.cc:
```diff
@@ -581,20 +590,37 @@ controller_backend::calculate_learner_initial_offset(
+          max_removable_local_log_offset,
+          last_uploaded,
+          last_uploaded);
+        max_removable_local_log_offset = last_uploaded;
```
this should be `std::min(stm.get_last_offset(), stm.get_last_clean_at());`

Lazin replied:
also, the TS is disabled maybe it makes more sense to just copy the log as if there is no TS to begin with

Lazin commented on src/v/cluster/controller_backend.cc:
```diff
@@ -581,20 +590,37 @@ controller_backend::calculate_learner_initial_offset(
+          last_uploaded);
+        max_removable_local_log_offset = last_uploaded;
+    }
+
```
maybe it makes sense to add `vassert(max_removable_local_log_offset != model::offset::max(), "....");` here

oleiman committed:
c/a: Introduce archival_metadata_stm::cloud_recoverable_offset

Returns the maximum offset which is guaranteed to be recoverable from
cloud storage.

This is the lesser of the last offset uploaded to cloud storage and the
last offset we uploaded a manifest for.

If the manifest is empty or the last uploaded offset is 0, returns
offset::min(), indicating that no data is recoverable from cloud.

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/e287125086ed4c7c75ed0e2648305f088c4268e9..9f2abd3512440420be26baaa66fcf0e2e07491f8) changes as discussed earlier

andrwng commented on src/v/cluster/archival/archival_metadata_stm.cc:
```diff
@@ -1306,6 +1306,21 @@ archival_metadata_stm::take_local_snapshot(ssx::semaphore_units apply_units) {
       0, snapshot_offset, std::move(snap_data));
 }
 
+model::offset archival_metadata_stm::max_safe_removable_offset() {
```
nit: suggestion: maybe `cloud_recoverable_offset()` or somesuch?

"removable" feels like it leaks out details of STM implementation / expectation of the caller, rather than describing what this is

oleiman replied:
yes, much better

oleiman replied:
done

andrwng commented on src/v/cluster/archival/archival_metadata_stm.h:
```diff
@@ -277,6 +277,16 @@ class archival_metadata_stm final : public raft::persisted_stm<> {
+    /// last offset we uploaded a manifest for.
+    ///
+    /// If the manifest is empty or the last uploaded offset is 0, returns
+    /// offset::min()
```
nit: ... offset::min(), indicating that nothing is recoverable from cloud.

oleiman replied:
factual

oleiman replied:
done

bharathv commented on src/v/cluster/controller_backend.cc:
```diff
@@ -496,12 +496,39 @@ controller_backend::calculate_learner_initial_offset(
      * storage data
      */
+    if (auto topic_cfg = p->get_topic_config();
+        topic_cfg.has_value() && topic_cfg->get().is_internal()) {
```
nit: I was thinking  is_user_topic(topic_namespace_view tp_ns) (seems a bit more stricter than is_internal).

oleiman replied:
yeah good call.

oleiman replied:
done

andrwng replied:
Isn't the behavior we're trying to capture tied to whether TS is enabled? There are non-user topics (schemas, audit, transforms) that correctly have TS enabled. Are we sure we want to change this behavior for these topics?

oleiman replied:
good spot. reverted in https://github.com/redpanda-data/redpanda/compare/26b990a7881ef50c601ca489efd3f170a4fbfb5e..3dc8d349563a192d63d2cfb16df93ebc4d3d03de

Lazin replied:
this is the check that defines what topics can enable TS - https://github.com/redpanda-data/redpanda/blob/ae7df7935e6b78dc597bddb7c9fc74c51fed846f/src/v/cluster/archival/archival_metadata_stm.cc#L1692

the requirement is that it's in kafka namespace and it's not a consumer offsets topic (the last bit is not backported, in old branches it's just a namespace check)

bharathv commented on src/v/cluster/controller_backend.cc:
```diff
@@ -496,12 +496,39 @@ controller_backend::calculate_learner_initial_offset(
         return std::nullopt;
     }
 
+    if (p->get_cloud_storage_mode() != cluster::cloud_storage_mode::full) {
```
do we still need the cloud_data_available()? (I think that that adds to confusion)

oleiman replied:
not 100% sure, but this comment (which I don't really understand) gives me pause:

```
    // If the start offset for the STM region is not set, then the cloud log is
    // empty. There's one special case, where the start offset is set, and yet
    // the cloud log should be considered emtpy: retention in the STM region
    // advanced the start offset, but the garbage collection, and subsequent
    // truncation, did not happen yet.
```

bharathv replied:
hmm ya, ok to skip then just to be on the safer side.

bharathv reviewed (commented):
couple of nits lgtm otherwise.

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-26803-v24.2.x-574 remotes/upstream/v24.2.x
git cherry-pick -x 668db629ae 3dc8d34956
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/16280078289).

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-26803-v24.3.x-103 remotes/upstream/v24.3.x
git cherry-pick -x 668db629ae 3dc8d34956
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/16280077441).
